tensorflow/core/kernels/concat_op.cc

```diff
@@ -201,21 +201,26 @@ TF_CALL_int64(REGISTER_GPU);
 // A special GPU kernel for int32.
 // TODO(b/25387198): Also enable int32 in device memory. This kernel
 // registration requires all int32 inputs and outputs to be in host memory.
-REGISTER_KERNEL_BUILDER(Name("Concat")
-                            .Device(DEVICE_GPU)
-                            .TypeConstraint<int32>("T")
-                            .HostMemory("concat_dim")
-                            .HostMemory("values")
-                            .HostMemory("output"),
-                        ConcatOp<CPUDevice, int32>);
-REGISTER_KERNEL_BUILDER(Name("ConcatV2")
-                            .Device(DEVICE_GPU)
-                            .TypeConstraint<int32>("T")
-                            .TypeConstraint<int32>("Tidx")
-                            .HostMemory("values")
-                            .HostMemory("axis")
-                            .HostMemory("output"),
-                        ConcatV2Op<CPUDevice, int32>);
+#define REGISTER_GPU(type)                                   \
+  REGISTER_KERNEL_BUILDER(Name("Concat")                     \
+                              .Device(DEVICE_GPU)            \
+                              .TypeConstraint<type>("T")     \
+                              .HostMemory("concat_dim")      \
+                              .HostMemory("values")          \
+                              .HostMemory("output"),         \
+                          ConcatOp<CPUDevice, type>)         \
+  REGISTER_KERNEL_BUILDER(Name("ConcatV2")                   \
```

Comment:
You don't want to register for the GPU like this, int32 is a weird special case that will hopefully someday go away.  You want to register in concat_lib_gpu.cc so that the operation actually runs on the GPU.